File: 3327-check-if-dfs-strings-are-palindromes/3327-check-if-dfs-strings-are-palindromes.cpp
#include <vector>
#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<bool> findAnswer(vector<int>& parent, string s) {
        int n = parent.size();
        vector<vector<int>> tree(n);
        vector<int> flarquintz = parent; 
        for (int i = 1; i < n; ++i) {
            tree[parent[i]].push_back(i);
        }

        vector<int> traversal;
        vector<int> start(n), end(n);
        dfs(0, tree, traversal, start, end);

        string traversal_string;
        for (int idx : traversal) {
            traversal_string += s[idx];
        }


        const int mod1 = 1e9 + 7, mod2 = 1e9 + 9;
        const int base1 = 911, base2 = 1597;
        vector<long long> pow_base1(n + 1), pow_base2(n + 1);
        pow_base1[0] = pow_base2[0] = 1;
        for (int i = 1; i <= n; ++i) {
            pow_base1[i] = (pow_base1[i - 1] * base1) % mod1;
            pow_base2[i] = (pow_base2[i - 1] * base2) % mod2;
        }


        vector<long long> prefix_hash1(n), prefix_hash2(n);
        prefix_hash1[0] = traversal_string[0];
        prefix_hash2[0] = traversal_string[0];
        for (int i = 1; i < n; ++i) {
            prefix_hash1[i] = (prefix_hash1[i - 1] * base1 + traversal_string[i]) % mod1;
            prefix_hash2[i] = (prefix_hash2[i - 1] * base2 + traversal_string[i]) % mod2;
        }


        vector<long long> suffix_hash1(n + 1), suffix_hash2(n + 1);
        suffix_hash1[n] = suffix_hash2[n] = 0;
        for (int i = n - 1; i >= 0; --i) {
            suffix_hash1[i] = (suffix_hash1[i + 1] * base1 + traversal_string[i]) % mod1;
            suffix_hash2[i] = (suffix_hash2[i + 1] * base2 + traversal_string[i]) % mod2;
        }

        vector<bool> answer(n);
        for (int i = 0; i < n; ++i) {
            int l = start[i], r = end[i];
            long long h1_fwd = get_hash(prefix_hash1, pow_base1, l, r, mod1);
            long long h2_fwd = get_hash(prefix_hash2, pow_base2, l, r, mod2);
            long long h1_rev = get_reverse_hash(suffix_hash1, pow_base1, l, r, mod1);
            long long h2_rev = get_reverse_hash(suffix_hash2, pow_base2, l, r, mod2);
            if (h1_fwd == h1_rev && h2_fwd == h2_rev) {
                answer[i] = true;
            } else {
                answer[i] = false;
            }
        }

        return answer;
    }

private:
    void dfs(int node, const vector<vector<int>>& tree, vector<int>& traversal, vector<int>& start, vector<int>& end) {
        start[node] = traversal.size();
        for (int child : tree[node]) {
            dfs(child, tree, traversal, start, end);
        }
        traversal.push_back(node);
        end[node] = traversal.size() - 1;
    }

    long long get_hash(const vector<long long>& prefix_hash, const vector<long long>& pow_base, int l, int r, int mod) {
        if (l == 0) return prefix_hash[r];
        return (prefix_hash[r] - prefix_hash[l - 1] * pow_base[r - l + 1] % mod + mod) % mod;
    }

    long long get_reverse_hash(const vector<long long>& suffix_hash, const vector<long long>& pow_base, int l, int r, int mod) {
        return (suffix_hash[l] - suffix_hash[r + 1] * pow_base[r - l + 1] % mod + mod) % mod;
    }
};

File: 3327-check-if-dfs-strings-are-palindromes/3327-check-if-dfs-strings-are-palindromes.py
class Solution:
    def findAnswer(self, parent: List[int], S: str) -> List[bool]:
        n = len(S)
        mod1, mod2 = 10**9 + 7, 10**9 + 9
        B1, B2 = 911, 3571
        max_len = n + 1

        pow_B1 = [1] * max_len
        pow_B2 = [1] * max_len
        for i in range(1, max_len):
            pow_B1[i] = (pow_B1[i - 1] * B1) % mod1
            pow_B2[i] = (pow_B2[i - 1] * B2) % mod2

        children = [[] for _ in range(n)]
        root = -1
        for i in range(n):
            if parent[i] == -1:
                root = i
            else:
                children[parent[i]].append(i)

        len_subtree = [0] * n
        H1 = [0] * n
        H2 = [0] * n
        HR1 = [0] * n
        HR2 = [0] * n
        answer = [False] * n

        def dfs(u):
            len_subtree[u] = 1
            H1_u, H2_u = 0, 0
            HR1_u = HR2_u = ord(S[u])
            for c in children[u]:
                dfs(c)
            for c in children[u]:
                H1_u = (H1_u * pow_B1[len_subtree[c]] + H1[c]) % mod1
                H2_u = (H2_u * pow_B2[len_subtree[c]] + H2[c]) % mod2
                len_subtree[u] += len_subtree[c]
            H1_u = (H1_u * B1 + ord(S[u])) % mod1
            H2_u = (H2_u * B2 + ord(S[u])) % mod2
            H1[u], H2[u] = H1_u, H2_u
            for c in reversed(children[u]):
                HR1_u = (HR1_u * pow_B1[len_subtree[c]] + HR1[c]) % mod1
                HR2_u = (HR2_u * pow_B2[len_subtree[c]] + HR2[c]) % mod2
            HR1[u], HR2[u] = HR1_u, HR2_u
            answer[u] = (H1[u] == HR1[u] and H2[u] == HR2[u])

        dfs(root)
        return answer